# dnasr - Reinterpretation of the SHAZAM algorithm for DNA

## Subcommands

Dnasr generate, produces database or query files from FASTA or FASTQ input. Raw Entropy information available with ` -- ee true`.

> Desktop/rust_projects/dnasr(master)» dnasr generate --help                                                                
> dnasr-generate 
> Generate queries and databases.
> 
> USAGE:
>     dnasr generate [OPTIONS] --db <BOOL> -i <FILE> -o <DIR>
> 
> FLAGS:  
>     -h, --help       Prints help information  
>     -V, --version    Prints version information  
> 
> OPTIONS:  
>         --db <BOOL>        Generate output as database or not.  
>         --ee <BOOL>        Subcommand generate should exit early and give raw information entropy. [default: false]  
>     -i <FILE>              FASTA or FASTQ file with sequences.  
>     -k <INT>               Kmer sizes to use. [default: 3]  
>         --mat <INT>        Columns in matrix. [default: 100]  
>        --np <INT>         Number of processes to use. [default: 1]  
>         --off <INT>        Offset between anchor and target zone. [default: 3]  
>     -o                Output directory. 
>         --sub_mat <INT>    Dimension of quadratic sub matrices. [default: 4]  
>     -w <INT>               Size of scanning window. [default: 150]  

dnasr search **not implemented yet**

> Desktop/rust_projects/dnasr(master)» dnasr search --help                                                                  
> dnasr-search 
> Search passed database with passed query.
> 
> USAGE:  
>     dnasr search --db <FILE> -o <DIR> -q <FILE>  
>
> FLAGS:  
>     -h, --help       Prints help information  
>     -V, --version    Prints version information  
> 
> OPTIONS:  
>         --db <FILE>    Database to search with passed query.  
>     -o                Output directory.  
>     -q <FILE>          Input file with query. Query generated by "dnas generate".   

## FAQ 
### How is this all working ?  
Please look at the [PDF](https://github.com/Luk13Mad/dnasr/tree/master/data) and the referenced literature within.
For more verbose code there is a [prototype](https://github.com/Luk13Mad/DNA-SHAZAM) written in python.

### And the short version of how this works?
The input DNA sequence gets split into kmers.
For each kmer the information entropy will be calculated from two very important properties of the kmer, the position and the number of occurences.
From there it is basically analog to the fingerprinting process also found in SHAZAM, but the frequency values will be replaced by the entropy and the time will be replaced by the position.  

### Cool. How can i get this to run on my machine?
Install Rust by following the [instructions](https://www.rust-lang.org/tools/install), clone this repository and use ``cargo run ``.

### Why did you do this?
This was part of my master thesis. The hoped results were not achieved but i thought it was a neat concept.

### Why is there an early exit option?
The SHAZAM finger prints are not very useful for further processing outside the SHAZAM environment. But the entropy information may be used further. **Following up on this soon**
  
## Performance
The data used for the figure can be found in this repository.
![performance](https://github.com/Luk13Mad/dnasr/blob/master/data/fig1.png)
